### ISSUE TITLE
Requests should not be sent to pod that just restarted

### ISSUE THREAD

**Issue by life1347** [fission/fission#1501]
A user reports that once a function pod (belongs to a function with executor type newdeploy) crashes and restarted, the sub-sequent requests sent to the same pod will get `Generic container: no requests supported` which means the function pod is not specialized yet. We should improve this in some ways.

**Comment by mattroark** [comment]
I've encountered this with the `poolmgr` executor as well on Fission v1.8.0 with my Golang environment.

```
❯ kubectl get po -n cs -L functionName
NAME                                   READY   STATUS        RESTARTS   AGE     FUNCTIONNAME
go-409854-75c5df7468-554cb             2/2     Running       2          4d21h
poolmgr-go-cs-409854-d47b6f8c9-qgqjb   2/2     Running       0          5m55s
poolmgr-go-cs-409854-d47b6f8c9-x8rp4   2/2     Terminating   1          8m52s   go-app
```

**Comment by soamvasani** [comment]
This looks like a bug in our readiness probe?  Since fetcher now controls the full specialization flow, it has enough information to be able to know when a pod is ready.

**Comment by life1347** [comment]
The fetcher mark readiness as ready after first time specialization, but it doesn't know the runtime container is crashed and restarted. Hence, the fetcher still marks readiness probe as true and wont call the specialization endpoint of runtime container.

**Comment by soamvasani** [comment]
Ah, thanks, I see.  

I'm thinking we could solve this in router/executor -- on getting the "generic container" error the router can switch to the cold start code path, call executor to specialize that container (or some other container), and then forward the request.

**Comment by life1347** [comment]
Well, I don’t think its a good idea and it didn’t solve the problem from the fundamental.

1. It increases the complexity of the call path
2. For newdeploy you cannot choose which pod router/executor talks with.
3. We are not able to control what custom env will returns.

**Comment by soamvasani** [comment]
Ok, radically simpler idea: set [restart policy to `never`](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy).

Then when the container crashes we simply discard the whole pod.  Everything else should work -- router already handles failures of specialized pods by re-attempting a cold start.

**Comment by life1347** [comment]
Does the Kubernetes discard the pod automatically or we have to decide which pod to delete? 

From the doc, 
```
Pod is running and has two Containers. Container 1 exits with failure.
If restartPolicy is: 
  Never: Do not restart Container; Pod phase stays Running.
```
So the Pod phase will stay in Running and the fetcher is still running which means the readiness probe returns true as well. Since the readiness probe is on fetcher, the Kubernetes still allow traffic to the pod. 

> router already handles failures of specialized pods by re-attempting a cold start

It's easier for the pool manager but tricky for newdeploy.

**Comment by life1347** [comment]
To be honest, I cannot think of a better way to solve this problem than #1536  (except for the upgrade issue).

**Comment by soamvasani** [comment]
> Since the readiness probe is on fetcher, the Kubernetes still allow traffic to the pod.

Yeah ok, that would have been too easy :)

#1536 has a bunch of pretty serious questions which I don't see any answers for, but let's continue that discussion in that issue.

> For newdeploy you cannot choose which pod router/executor talks with.

This is a good point.  Ok, let me try again:

 1. Router routes a request 
 1. Runtime environment crashes
 1. That request errors out
 1. Future requests get sent to the restarted version of that runtime
 1. Router gets a generic container error
 1. Router sends executor a new kind of request -- "I found a bad pod, please destroy it".  In poolmgr, this is easy to implement; in newdeploy it's not easy to find the bad pod, but I think we could do it by destroying any pod that has restarts > 0.
 1. Executor destroys the bad pod
 1. Router re-attempts the request
 1. This time it succeeds

**Comment by matthieu-eck** [comment]
Hi, I would like to know if there is an existing workaround for this issue,
Thanks!

**Comment by sanketsudake** [comment]
There is workaround implemented in the python environment for this issue.

To avoid containers not serving after OOM restart I would recommend this fix which is implemented in the Fission Python environment.

https://github.com/fission/environments/pull/213

**Comment by airhorns** [comment]
Can anyone say if this is still Fission's behaviour to send requests to pods that have restarted?

**Comment by sanketsudake** [comment]
Addressed on `main` by the RFC-0002 EndpointSlice data plane: the router's endpoint index admits only `Ready` EndpointSlice endpoints, so a pod that just restarted (`Ready=false`) is not sent traffic, and both the cold-start `choosePod` and warm-admit paths gate on pod readiness (`pkg/router/endpointcache/index.go`, `pkg/utils/podutils.go`).

Closing as resolved — reopen if you still observe requests hitting an unready pod on a current build.
